src/ja/server/database/types/work_machine.py

```diff
@@ -0,0 +1,95 @@
+from ja.common.work_machine import ResourceAllocation
+
+
+class RunningWorkMachineResources:
+    """
+    Represents the resources of a work machine which is currently connected to
+    the central server.
+
+    In addition to the available resources, a running work machine also tracks
+    how much of its resources are currently used.
+    """
+    def __init__(self, total_resources: ResourceAllocation):
+        """!
+        Initialize the RunningWorkMachineResources object. In the beginning,
+        all resources are marked as free.
+
+        @param total_resources The total amount of available resources.
+        """
+
+    @property
+    def total_resources(self) -> ResourceAllocation:
+        """!
+        @return The maximum resources available on the work machine.
+        """
+
+    @property
+    def free_resources(self) -> ResourceAllocation:
+        """!
+        @return The currently available resources on the work machine.
+        """
+
+    def allocate(self,
+                 allocation: ResourceAllocation,
+                 test_only: bool = False) -> bool:
+        """!
+        Add the requested resources to the list of already allocated resources.
+        These resources will not be marked as free afterwards.
+
+        @param allocation The CPU, memory and swap space to allocate.
+        @param test_only If set to True, only the availability of enough
+          resources will be checked, but the resources will not be actually
+          allocated.
+        @return True if the allocation or the test was successful, False
+          otherwise.
+        """
+
```

Comment:
There should be a method that marks the resources as free, for example after a Job finishes we should mark the resources as free in the database.